### PR TITLE
Derive Qwen3-TTS codec conv encoder from config instead of hardcoding widths

### PR DESCRIPTION
**Pull request by justinchuby**
## The bug

`_MimiConvEncoder` in `src/mobius/models/qwen3_tts_tokenizer.py` took **no config at all** and hardcoded its channel widths and kernels (`1→64→128→256→512→1024→512`). Its output fed `CodecEncoderTransformerModel(hidden_size=config.codec_encoder.hidden_size)`, which has no input projection — so the conv output width and `hidden_size` had to agree, and nothing checked that they did.

When they disagreed, the builder happily emitted a malformed graph:

```
Conv encoder/encoder/layers/14/conv  weight [512,1024,3]  -> [batch, T, 512]
LayerNormalization  weight [32] bias [32]  over a 512-wide tensor   <-- malformed
MatMul  q_proj [32,32] vs 512 -> "Incompatible MatMul contraction dimensions: 512 vs 32"
```

## Why it was invisible

`SymbolicShapeInferencePass` catches inference errors and logs `"Symbolic shape inference failed (upstream bug); skipping"` — a deliberate generic guard. The tiny config in `TestBuildCodecGraph._codec_config()` set `codec_encoder.hidden_size=32`, so **the codec tests were building a broken graph and passing**.

## The fix

The hardcoded widths were not dictated by the checkpoint — HF's `MimiEncoder.__init__` derives the entire stack from config, and mobius's literals were just that derivation inlined with `num_filters=64`, `upsampling_ratios=[8,6,5,4]`, `num_residual_layers=1`, `kernel_size=7`, `last_kernel_size=3`. So rather than adding a validation guard, this parameterizes the stack and eliminates the mismatch class:

- **`CodecEncoderConfig`**: added the missing HF fields (`audio_channels`, `num_filters`, `num_residual_layers`, `kernel_size`, `last_kernel_size`, `residual_kernel_size`, `compress`, `upsampling_ratios`) with HF's own defaults, extracted in `ArchitectureConfig.from_transformers`.
- **`_MimiConvEncoder`**: builds by HF's derivation. The parameterless ELU slots are preserved so `encoder.encoder.layers.*` numbering still matches the checkpoint exactly. `kernel = ratio * 2` and the trailing conv's out-channels is `hidden_size`, both derived.
- **`_EncoderResBlock`**: inner width is now `dim // compress` with a configurable kernel.

## Also fixed: the encoder RVQ could not load the real checkpoint

This one is a **weight-loading correctness fix, not just a shape-consistency fix**, which is why it belongs here despite being adjacent to the original scope.

`_EncoderSplitRVQ` used `input_dim=codebook_dim` and `dim=codebook_dim // 2`. The real checkpoint has `encoder.quantizer.*.input_proj.weight [256, 512, 1]` and `codebook.embed_sum [2048, 256]`, with `encoder_config.codebook_dim=256`. So the old code built `Conv1d(256→128)` projections and 128-wide codebooks: **the encoder could never have loaded the real checkpoint's weights**, independently of the malformed-graph issue. It only happened to be self-consistent on the `enc is None` fallback path, where `codebook_dim` defaulted to 512.

Now it projects `hidden_size → codebook_dim`, matching HF `MimiResidualVectorQuantizer`.

### The encoder/decoder asymmetry is intentional — please don't unify these paths

The two sides interpret `codebook_dim` differently, and both are correct:

| | config value | RVQ input | codebook dim | checkpoint `input_proj` |
|---|---|---|---|---|
| encoder | `codebook_dim=256` (Mimi semantics: the codebook dim itself) | `hidden_size=512` | 256 | `[256, 512, 1]` |
| decoder | `codebook_dim=512` (the RVQ *input* dim) | 512 | `512 // 2 = 256` | `[256, 512, 1]` |

So the decoder's `// 2` convention in `SplitResidualVectorQuantizer` is right and is left untouched; only the encoder path was misreading its field. A future "cleanup" that unifies the two would reintroduce this bug.

## Verification

**Symbolic-shape-inference failures across `TestBuildCodecGraph`** (measured by temporarily instrumenting `infer_symbolic_shapes`, since the pass swallows them — instrumentation not committed):

| | failures |
|---|---|
| before | **6** |
| after | **0** |

**Default config reproduces the real checkpoint exactly** — verified against the safetensors header of `Qwen/Qwen3-TTS-Tokenizer-12Hz`: all 14 conv-stack weights match by name and shape, including the ResBlock `layers.N.block.1/.3` names and the `0,1,3,4,6,7,9,10,12,14` numbering, as do both RVQ `input_proj`/`output_proj` pairs and the codebooks. This is asserted by tests, so indexing or width drift (which would otherwise surface far away at weight-load time) fails loudly.

**Tiny test config** keeps the real depth and only shrinks widths (`num_filters=4`, `hidden_size=32` → `1→4→8→16→32→64→32`), so it exercises the same weight-name structure the real model depends on. Codec test runtime: 2.05s (4 tests) → ~2.1s (9 tests).

**Suite**: `pytest tests/build_graph_test.py tests/cli_test.py src/ -k "not phi4mm and not apply_weights_unknown" -n auto` → 8 failed, 4613 passed — the same 8 pre-existing failures on `main` (onnx_genai jsonschema drift + one qwen_image golden). `lintrunner` clean.

## New tests

- `test_default_config_conv_stack_matches_checkpoint` — asserts derived parameter **names** and shapes against the real checkpoint.
- `test_tiny_config_keeps_checkpoint_layer_names` — tiny config has identical naming, narrower tensors.
- `test_conv_stack_indices_follow_config` — 2 ratios → final conv at `layers.8`; 4 ratios × 2 residual layers → `layers.18`.
- `test_hidden_size_drives_conv_output_width` — direct regression guard for the original bug.
- `test_default_config_rvq_projections_match_checkpoint` — asserts `input_proj (256, 512, 1)`, `output_proj (512, 256, 1)` and `(2048, 256)` codebooks, guarding the weight-loading fix.

## Decoder conv stack

Checked: the codec **decoder** does **not** have the mirrored conv problem. `Qwen3TTSCodecDecoderModel` already derives everything from config (`latent_dim`, `decoder_dim`, `upsample_rates`, `upsampling_ratios`), and `CodecDecoderTransformerModel` has an `input_proj`/`output_proj` pair that adapts `latent_dim ↔ hidden_size`, so no width can be stranded. No change needed there.

## Scope

Limited to the codec tokenizer model, its config, and its tests. Does not touch `reuse.py`, `_ssm.py`, `_mamba_block.py`, the onnx_genai integration (covered by #551), or `SymbolicShapeInferencePass`.

### CODE REVIEW

File: src/mobius/_configs/_base.py [hunk 1]
```diff
@@ -1125,6 +1125,14 @@ def from_transformers(cls, config, parent_config=None) -> ArchitectureConfig:
                 max_position_embeddings=getattr(ec, "max_position_embeddings", 8000),
                 num_quantizers=getattr(ec, "num_quantizers", 32),
                 num_semantic_quantizers=getattr(ec, "num_semantic_quantizers", 1),
+                audio_channels=getattr(ec, "audio_channels", 1),
+                num_filters=getattr(ec, "num_filters", 64),
+                num_residual_layers=getattr(ec, "num_residual_layers", 1),
+                kernel_size=getattr(ec, "kernel_size", 7),
+                last_kernel_size=getattr(ec, "last_kernel_size", 3),
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/_configs/_base.py:1132]: Good catch — the existing tests construct `CodecEncoderConfig` directly, so a wrong `getattr` key or default here would have passed silently.

Added three extraction tests in `src/mobius/_configs/_base_test.py`, all driving `ArchitectureConfig.from_transformers` off a nested HF-style config:

- `test_codec_encoder_conv_fields_extracted_from_nested_config` — checkpoint values round-trip.
- `test_codec_encoder_conv_fields_honor_non_default_values` — non-default values for every new field, including `upsampling_ratios=[4, 2]` and `audio_channels=2`, survive extraction unchanged. This is the one that would fail on a mistyped key.
- `test_codec_encoder_conv_fields_fall_back_to_checkpoint_defaults` — a config omitting the conv fields still yields the real architecture (`num_filters=64`, `upsampling_ratios=[8, 6, 5, 4]`, `num_residual_layers=1`), so older configs keep working.

File: src/mobius/models/qwen3_tts_tokenizer.py [hunk 2]
```diff
@@ -456,28 +505,32 @@ class _EncoderSplitRVQ(nn.Module):
...
 
         # Semantic quantizer
         self.semantic_residual_vector_quantizer = _EncoderRVQ(
             num_quantizers=num_semantic,
             codebook_size=codebook_size,
             dim=dim,
-            input_dim=codebook_dim,
-            output_dim=codebook_dim,
+            input_dim=hidden_size,
+            output_dim=hidden_size,
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/qwen3_tts_tokenizer.py:525]: This one is already covered — `test_default_config_rvq_projections_match_checkpoint` landed in 3b3f84b, which is the head this review ran against, so I think it was just missed.

It asserts, for the default config, that both the semantic and acoustic quantizers have `input_proj == (256, 512, 1)` (hidden_size → codebook_dim) and `output_proj == (512, 256, 1)` (the reverse), plus `(2048, 256)` codebooks — matching `Qwen/Qwen3-TTS-Tokenizer-12Hz`'s `encoder.quantizer.*` weights exactly. Since these are checked as parameter shapes rather than via inference, the swallowed-symbolic-error problem you're pointing at doesn't apply.

Agreed on the underlying reasoning though: this was a weight-loading bug, not just a shape-consistency one. Before the fix the encoder built `Conv1d(256→128)` projections and 128-wide codebooks, so it could never have loaded the real checkpoint at all.

File: src/mobius/models/qwen3_tts_tokenizer.py [hunk 3]
```diff
@@ -223,20 +225,22 @@ def __init__(self, config: ArchitectureConfig):
...
-        #  14: Conv1d(1024->512, k=3)
-        self.encoder = _MimiConvEncoder()
+        # Conv encoder: series of Conv1d + residual blocks, derived from
+        # config exactly like HF ``MimiEncoder.__init__``. With the
+        # checkpoint's defaults (num_filters=64, upsampling_ratios=[8,6,5,4],
+        # num_residual_layers=1, kernel_size=7, last_kernel_size=3) this
+        # yields 1->64->128->256->512->1024->hidden_size.
+        self.encoder = _MimiConvEncoder(
+            hidden_size=hidden_size,
+            audio_channels=enc.audio_channels if enc else 1,
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/qwen3_tts_tokenizer.py:235]: Correct, and thanks — this was a real gap I introduced. Making `audio_channels` configurable without threading it through left `CodecTask._build_encoder` declaring `[batch, 1, audio_length]` against a conv sized for `audio_channels`.

Fixed by threading it into the task input shape rather than rejecting non-mono, since the field is genuinely part of the Mimi architecture:

```python
audio_channels = config.codec_encoder.audio_channels if config.codec_encoder else 1
waveform = builder.input(
    "waveform", dtype=ir.DataType.FLOAT, shape=[batch, audio_channels, audio_len]
)
```

The `None` fallback keeps Mimi (`mimi_default_config()` carries no `codec_encoder`) on a mono input, and its tests still pass.

Added `test_encoder_input_declares_configured_audio_channels`, which builds the full package with `audio_channels=2` and asserts the graph input's channel dim is 2 — that build would have produced the mismatched graph before this fix.
